### PR TITLE
Fix --base-image caching for images specified by name:tag

### PR DESCRIPTION
**Pull request by ilya-zuyev**
This PR fixes an issue when the image set by `--base-image` is downloaded on start even if it's already present in minikube cache. 
### Before:
```
➜  minikube git:(master) ✗ time m start  --base-image=docker.io/ilyazuyev/kicbase:jun4_0 
😄  minikube v1.21.0-beta.0 on Darwin 11.3.1
✨  Automatically selected the docker driver. Other choices: virtualbox, ssh
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
    > index.docker.io/ilyazuyev/k...: 359.08 MiB / 359.08 MiB  100.00% 5.63 MiB
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.20.7 on Docker 20.10.7 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
out/minikube start --base-image=docker.io/ilyazuyev/kicbase:jun4_0  11.13s user 7.60s system 13% cpu 2:23.21 total

➜  minikube git:(master) ✗ m delete --all
🔥  Deleting "minikube" in docker ...
🔥  Removing /Users/izuyev/.minikube/machines/minikube ...
💀  Removed all traces of the "minikube" cluster.
🔥  Successfully deleted all profiles

➜  minikube git:(master) ✗ time m start  --base-image=docker.io/ilyazuyev/kicbase:jun4_0 
😄  minikube v1.21.0-beta.0 on Darwin 11.3.1
✨  Automatically selected the docker driver. Other choices: virtualbox, ssh
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
    > index.docker.io/ilyazuyev/k...: 359.08 MiB / 359.08 MiB  100.00% 5.73 MiB
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.20.7 on Docker 20.10.7 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
out/minikube start --base-image=docker.io/ilyazuyev/kicbase:jun4_0  13.43s user 10.21s system 15% cpu 2:36.20 total

```
### After:
```
➜  minikube git:(ilyaz/fix_no_digest_caching) ✗ time m start  --base-image=docker.io/ilyazuyev/kicbase:jun4_0                              
😄  minikube v1.21.0-beta.0 on Darwin 11.3.1
✨  Automatically selected the docker driver. Other choices: virtualbox, ssh
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
💾  Downloading Kubernetes v1.20.7 preload ...
    > preloaded-images-k8s-v11-v1...: 492.20 MiB / 492.20 MiB  100.00% 6.57 MiB
    > index.docker.io/ilyazuyev/k...: 359.08 MiB / 359.08 MiB  100.00% 2.91 MiB
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.20.7 on Docker 20.10.7 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
out/minikube start --base-image=docker.io/ilyazuyev/kicbase:jun4_0  16.40s user 13.19s system 13% cpu 3:41.00 total


➜  minikube git:(ilyaz/fix_no_digest_caching) ✗ m delete --all
🔥  Deleting "minikube" in docker ...
🔥  Removing /Users/izuyev/.minikube/machines/minikube ...
💀  Removed all traces of the "minikube" cluster.
🔥  Successfully deleted all profiles


➜  minikube git:(ilyaz/fix_no_digest_caching) ✗ time m start  --base-image=docker.io/ilyazuyev/kicbase:jun4_0 
😄  minikube v1.21.0-beta.0 on Darwin 11.3.1
✨  Automatically selected the docker driver. Other choices: virtualbox, ssh
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.20.7 on Docker 20.10.7 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
out/minikube start --base-image=docker.io/ilyazuyev/kicbase:jun4_0  9.32s user 5.43s system 15% cpu 1:36.86 total

```
Note, after the fix, the base image is *not* downloaded.

There is an exception for `latest`, when `--base-image=:latest` format is used the image is always pulled

### CODE REVIEW

File: pkg/minikube/download/image.go [hunk 1]
```diff
@@ -168,17 +169,40 @@ func ImageToCache(img string) error {
...
+func parseImage(img string) (*name.Tag, name.Reference, error) {
+	digest, err := name.NewDigest(img)
+	if err == nil {
+		tag := digest.Tag()
+		return &tag, digest, nil
+	}
+
+	_, ok := err.(*name.ErrBadName)
+	if !ok {
+		return nil, nil, errors.Wrap(err, "new ref")
```
Comment (medyagh) [inline pkg/minikube/download/image.go:181]: change this to a better wrap
```
errors.Wrap(err, "bad name")
```

**Review by medyagh — CHANGES_REQUESTED** [review]
thank you for this PR this seems interesting !

File: pkg/minikube/download/image.go [hunk 2]
```diff
@@ -168,17 +169,40 @@ func ImageToCache(img string) error {
...
 }
 
+func parseImage(img string) (*name.Tag, name.Reference, error) {
+	digest, err := name.NewDigest(img)
+	if err == nil {
+		tag := digest.Tag()
+		return &tag, digest, nil
+	}
+
+	_, ok := err.(*name.ErrBadName)
```
Comment (medyagh) [inline pkg/minikube/download/image.go:179]: why we are only checking for this specific errors ? 
add comment when does this specific error happen ?
seems like we are ignoring the error if it is not ErrBadName ?

**Review by medyagh — CHANGES_REQUESTED** [review]

**Comment by medyagh** [comment]
@ilya-zuyev please pull upstream on this PR, there was a another PR that touched same part of this code might need to test with that

File: pkg/minikube/download/image.go (hunk 2 above)
Comment (ilya-zuyev) [reply pkg/minikube/download/image.go:179]: Exactly. `ErrBadName` means `img` contains no digest. It happens if its value is `name:tag` for example. In this case we want to give it a second chance and try to parse it one more time using `name.NewTag(img)`

File: pkg/minikube/download/image.go (hunk 1 above)
Comment (ilya-zuyev) [reply pkg/minikube/download/image.go:181]: It's not a bad name. It's bad image reference

**Comment by ilya-zuyev** [comment]
> @ilya-zuyev please pull upstream on this PR, there was a another PR that touched same part of this code might need to test with that

done!

File: pkg/minikube/download/image.go (hunk 2 above)
Comment (medyagh) [reply pkg/minikube/download/image.go:179]: lets add this as a comment

File: pkg/minikube/download/image.go (hunk 1 above)
Comment (medyagh) [reply pkg/minikube/download/image.go:181]: then change to "bad ref" so it is different than the next one which is also "new ref"

**Review by medyagh — CHANGES_REQUESTED** [review]
lets fix the review comments and hopefully cloud shell test will pass this time too (seems like it didnt get started)

File: pkg/minikube/download/image.go (hunk 1 above)
Comment (ilya-zuyev) [reply pkg/minikube/download/image.go:181]: But for callers it's the same error - `img` contains no valid image reference. 
 "new ref" is a terrible error message, actually, but it's already in the code

File: pkg/minikube/download/image.go (hunk 2 above)
Comment (ilya-zuyev) [reply pkg/minikube/download/image.go:179]: done

File: pkg/minikube/download/image.go (hunk 1 above)
Comment (ilya-zuyev) [reply pkg/minikube/download/image.go:181]: changed

**Review by medyagh — APPROVED** [review]

**Pull request merged**